Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 78234b1aea62905fc445790cf91254e16716dbce
Number of Error(s): 4
/mnt/disk1/home/vamosax/smoketest/PR-12105/HPCC-Platform/esp/services/ws_configmgr/ws_configmgrService.cpp:26:58: fatal error: ../../../../release/generated/ws_configmgr.hpp: No such file or directory
make[2]: *** [esp/services/ws_configmgr/CMakeFiles/ws_configmgr.dir/ws_configmgrService.cpp.o] Error 1

[Build error:](https://gist.github.com/719914b861f2b2de153df124466b6d18)
Build log [Build log](https://gist.github.com/HPCCSmoketest/719914b861f2b2de153df124466b6d18#file-result-19-01-28-14-47-11-log)

--- a/configuration/configmgr/configmgrlib/EnvironmentLoader.hpp
+++ b/configuration/configmgr/configmgrlib/EnvironmentLoader.hpp
@@ -36,7 +36,8 @@ class CFGMGRLIB_API EnvironmentLoader
 
         EnvironmentLoader() { }
         virtual ~EnvironmentLoader() { }
-        virtual std::vector<std::shared_ptr<EnvironmentNode>> load(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem) const = 0;
+        virtual std::vector<std::shared_ptr<EnvironmentNode>> load(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem,
+                const std::string itemType = std::string("")) const = 0;
 
 
     protected:

--- a/configuration/configmgr/configmgrlib/EnvironmentMgr.cpp
+++ b/configuration/configmgr/configmgrlib/EnvironmentMgr.cpp
@@ -160,6 +160,13 @@ std::string EnvironmentMgr::getRootNodeId() const
 }
 
 
+bool EnvironmentMgr::serialize(std::ostream &out, const std::shared_ptr<EnvironmentNode> &pStartNode)
+{
+   return false;
+}
+
+
+
 bool EnvironmentMgr::saveEnvironment(const std::string &qualifiedFilename)
 {
     bool rc = false;
@@ -249,7 +256,7 @@ std::shared_ptr<EnvironmentNode> EnvironmentMgr::addNewEnvironmentNode(const std
             }
             else
             {
-                pNewNode->validate(status, true, false);
+                m_pSchema->validate(status, true, false);
             }
         }
         else
@@ -306,6 +313,27 @@ std::shared_ptr<EnvironmentNode> EnvironmentMgr::addNewEnvironmentNode(const std
         }
     }
 
+    return pNewEnvNode;
+}
+
+
+std::shared_ptr<EnvironmentNode> EnvironmentMgr::addNewEnvironmentNode(const std::shared_ptr<EnvironmentNode> &pParentNode, const std::string &nodeData,
+        Status &status, const std::string &itemType)
+{
+    std::shared_ptr<EnvironmentNode> pNewEnvNode;
+    std::istringstream newNodeData(nodeData);
+    std::vector<std::shared_ptr<EnvironmentNode>> newNodes = doLoadEnvironment(newNodeData, pParentNode->getSchemaItem(), itemType);  // not root
+    if (!newNodes.empty())
+    {
+        pNewEnvNode = newNodes[0];
+        pParentNode->addChild(pNewEnvNode);
+        assignNodeIds(pNewEnvNode);
+        validate(status);
+    }
+    else
+    {
+        status.addMsg(statusMsg::error, "Error pasting new node data.");
+    }
     return pNewEnvNode;
 }
 
@@ -410,7 +438,7 @@ void EnvironmentMgr::validate(Status &status, bool includeHiddenNodes) const
 {
     if (m_pRootNode)
     {
-        m_pRootNode->validate(status, true, includeHiddenNodes);
+        m_pSchema->validate(status, true, includeHiddenNodes);
     }
     else
     {

--- a/configuration/configmgr/configmgrlib/EnvironmentMgr.hpp
+++ b/configuration/configmgr/configmgrlib/EnvironmentMgr.hpp
@@ -57,9 +57,12 @@ class CFGMGRLIB_API EnvironmentMgr
         std::shared_ptr<EnvironmentNode> getNewEnvironmentNode(const std::string &parentNodeId, const std::string &inputItem, Status &status) const;
         std::shared_ptr<EnvironmentNode> addNewEnvironmentNode(const std::string &parentNodeId, const std::string &configType,
                 std::vector<NameValue> &initAttributes, Status &status, bool allowInvalid=false, bool forceCreate=false);
+        std::shared_ptr<EnvironmentNode> addNewEnvironmentNode(const std::shared_ptr<EnvironmentNode> &pParentNode, const std::string &nodeData,
+                Status &status, const std::string &itemType);
         std::shared_ptr<EnvironmentNode> addNewEnvironmentNode(const std::shared_ptr<EnvironmentNode> &pParentNode, const std::shared_ptr<SchemaItem> &pNewCfgItem,
                 std::vector<NameValue> &initAttributes, Status &status, bool allowInvalid=false, bool forceCreate=false);
         bool removeEnvironmentNode(const std::string &nodeId);
+        virtual bool serialize(std::ostream &out, const std::shared_ptr<EnvironmentNode> &pStartNode);
         bool saveEnvironment(const std::string &qualifiedFilename);
         void discardEnvironment() { m_pRootNode = nullptr; m_nodeIds.clear();}
         void validate(Status &status, bool includeHiddenNodes=false) const;
@@ -74,7 +77,7 @@ class CFGMGRLIB_API EnvironmentMgr
 
         void addPath(const std::shared_ptr<EnvironmentNode> pNode);
         virtual bool createParser() = 0;
-        virtual std::vector<std::shared_ptr<EnvironmentNode>> doLoadEnvironment(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem) = 0;
+        virtual std::vector<std::shared_ptr<EnvironmentNode>> doLoadEnvironment(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem, const std::string itemType = std::string("")) = 0;
         virtual bool save(std::ostream &out) = 0;
         void assignNodeIds(const std::shared_ptr<EnvironmentNode> &pNode);
         void insertExtraEnvironmentData(std::shared_ptr<EnvironmentNode> pNode);

--- a/configuration/configmgr/configmgrlib/EnvironmentNode.cpp
+++ b/configuration/configmgr/configmgrlib/EnvironmentNode.cpp
@@ -217,66 +217,53 @@ std::string EnvironmentNode::getLocalValue() const
 }
 
 
-void EnvironmentNode::validate(Status &status, bool includeChildren, bool includeHiddenNodes) const
+void EnvironmentNode::validate(Status &status) const
 {
-    if (!m_pSchemaItem->isHidden() || includeHiddenNodes)
+    //
+    // Check node value
+    if (m_pLocalValue)
     {
-        //
-        // Check node value
-        if (m_pLocalValue)
-        {
-            m_pLocalValue->validate(status, m_id);
-        }
+        m_pLocalValue->validate(status, m_id);
+    }
+
+    //
+    // Check any attributes
+    for (auto &attrIt: m_attributes)
+    {
+        attrIt.second->validate(status, m_id);
 
         //
-        // Check any attributes
-        for (auto attrIt = m_attributes.begin(); attrIt != m_attributes.end(); ++attrIt)
+        // If this value must be unique, make sure it is
+        if (attrIt.second->getSchemaValue()->isUniqueValue())
         {
-            attrIt->second->validate(status, m_id);
-
-            //
-            // If this value must be unique, make sure it is
-            if (attrIt->second->getSchemaValue()->isUniqueValue())
+            bool found = false;
+            std::vector<std::string> allValues;
+            attrIt.second->getAllValuesForSiblings(allValues);
+            std::set<std::string> unquieValues;
+            for (auto it = allValues.begin(); it != allValues.end() && !found; ++it)
             {
-                bool found = false;
-                std::vector<std::string> allValues;
-                attrIt->second->getAllValuesForSiblings(allValues);
-                std::set<std::string> unquieValues;
-                for (auto it = allValues.begin(); it != allValues.end() && !found; ++it)
-                {
-                    auto ret = unquieValues.insert(*it);
-                    found = !ret.second;
-                }
-
-                if (found)
-                {
-                    status.addUniqueMsg(statusMsg::error, m_id, attrIt->second->getName(), "Attribute value must be unique");
-                }
+                auto ret = unquieValues.insert(*it);
+                found = !ret.second;
             }
 
-            //
-            // Does this value need to be from another set of values?
-            if (attrIt->second->getSchemaValue()->isFromUniqueValueSet())
+            if (found)
             {
-                bool found = false;
-                std::vector<std::string> allValues;
-                attrIt->second->getSchemaValue()->getAllKeyRefValues(allValues);
-                for (auto it = allValues.begin(); it != allValues.end() && !found; ++it)
-                    found = *it == attrIt->second->getValue();
-                if (!found)
-                {
-                    status.addMsg(statusMsg::error, m_id, attrIt->second->getName(), "Attribute value must be from a unique set");
-                }
+                status.addUniqueMsg(statusMsg::error, m_id, attrIt.second->getName(), "Attribute value must be unique");
             }
         }
 
         //
-        // Now check all children
-        if (includeChildren)
+        // Does this value need to be from another set of values?
+        if (attrIt.second->getSchemaValue()->isFromUniqueValueSet())
         {
-            for (auto childIt = m_children.begin(); childIt != m_children.end(); ++childIt)
+            bool found = false;
+            std::vector<std::string> allValues;
+            attrIt.second->getSchemaValue()->getAllKeyRefValues(allValues);
+            for (auto it = allValues.begin(); it != allValues.end() && !found; ++it)
+                found = *it == attrIt.second->getValue();
+            if (!found)
             {
-                childIt->second->validate(status, includeChildren, includeHiddenNodes);
+                status.addMsg(statusMsg::error, m_id, attrIt.second->getName(), "Attribute value must be from a unique set");
             }
         }
     }

--- a/configuration/configmgr/configmgrlib/EnvironmentNode.hpp
+++ b/configuration/configmgr/configmgrlib/EnvironmentNode.hpp
@@ -63,7 +63,7 @@ class CFGMGRLIB_API EnvironmentNode : public std::enable_shared_from_this<Enviro
         bool hasAttributes() const { return m_attributes.size() != 0; }
         void setId(const std::string &id) { m_id = id; }
         const std::string &getId() const { return m_id;  }
-        void validate(Status &status, bool includeChildren=false, bool includeHiddenNodes=false) const;
+        void validate(Status &status) const;
         void getAttributeValueForAllSiblings(const std::string &attrName, std::vector<std::string> &result) const;
         const std::shared_ptr<SchemaItem> &getSchemaItem() const { return m_pSchemaItem; }
         void getInsertableItems(std::vector<InsertableItem> &items) const;

--- a/configuration/configmgr/configmgrlib/SchemaItem.cpp
+++ b/configuration/configmgr/configmgrlib/SchemaItem.cpp
@@ -21,6 +21,7 @@
 #include "SchemaTypeIntegerLimits.hpp"
 #include <algorithm>
 #include "Utils.hpp"
+#include "EnvironmentNode.hpp"
 
 
 SchemaItem::SchemaItem(const std::string &name, const std::string &className, const std::shared_ptr<SchemaItem> &pParent) :
@@ -474,9 +475,9 @@ std::shared_ptr<SchemaItem> SchemaItem::getSchemaRoot()
 void SchemaItem::getPath(std::string &path) const
 {
     path = getProperty("name") + path;
+    path = "/" + path;
     if (!m_pParent.expired())
     {
-        path = "/" + path;
         m_pParent.lock()->getPath(path);
     }
 }
@@ -645,4 +646,94 @@ void SchemaItem::processEvent(const std::string &eventType, const std::shared_pt
 void SchemaItem::setRequiredInstanceComponents(const std::string list)
 {
     m_requiredInstanceComponents = splitString(list, "|");
+}
+
+
+void SchemaItem::validate(Status &status, bool includeChildren, bool includeHiddenNodes) const
+{
+    //
+    // Check the number of environment nodes against the allowed min/max range. Note that since the schema does not
+    // maintain a parent child relationship like the environment, the environment nodes for this schema item need to
+    // be separated by each environment node's parent. Then the number of nodes, by parent, is compared agains the
+    // allowed min/max range
+    std::map<std::string, std::vector<std::shared_ptr<EnvironmentNode>>> envNodesByParent;
+
+    //
+    // Fill in the nodes by parent so that we know how many of this schema item are under each parent. Note some will
+    // be 0 and this needs to be known in case the min number is not 0. If there is not parent for this item, then we
+    // are at the root and no check is necessary (expired check OK since the schema is static once loaded)
+    if (!m_pParent.expired())
+    {
+        auto pParentItem = m_pParent.lock();  // also, schema items are not deleted, so this will always be good
+        for (auto &pParentWeakEnvNode: pParentItem->m_envNodes)
+        {
+            std::shared_ptr<EnvironmentNode> pParentEnvNode = pParentWeakEnvNode.lock();
+            if (pParentEnvNode)
+            {
+                envNodesByParent[pParentEnvNode->getId()] = std::vector<std::shared_ptr<EnvironmentNode>>();
+            }
+        }
+    }
+
+    //
+    // Now build the vector per parent node ID
+    for (auto &pWeakEnvNode: m_envNodes)
+    {
+        std::shared_ptr<EnvironmentNode> pEnvNode = pWeakEnvNode.lock();
+        if (pEnvNode)
+        {
+            std::shared_ptr<EnvironmentNode> pParentEnvNode = pEnvNode->getParent();
+            if (pParentEnvNode)
+            {
+                envNodesByParent[pParentEnvNode->getId()].emplace_back(pEnvNode);
+            }
+        }
+    }
+
+    //
+    // Now validate the count of nodes
+    for (auto &envNodes: envNodesByParent)
+    {
+        if (envNodes.second.size() < m_minInstances || envNodes.second.size() > m_maxInstances)
+        {
+            for (auto &pEnvNode: envNodes.second)
+            {
+                std::string path;
+                getPath(path);
+                std::string msg;
+                msg = "The number of SchemaItem " + path + " nodes (" + std::to_string(envNodes.second.size()) +
+                      ") is outside the range of " + std::to_string(m_minInstances) +
+                      " to " + std::to_string(m_maxInstances);
+                status.addMsg(statusMsg::error, pEnvNode->getId(), "", msg);
+            }
+        }
+    }
+
+
+    //
+    // Now validate each environment node (at this time go ahead and reset the member variable of weak pointers as well
+    // to get rid of any stale values)
+    for (auto &pWeakEnvNode: m_envNodes)
+    {
+        if (!isHidden() || includeHiddenNodes)
+        {
+            std::shared_ptr<EnvironmentNode> pEnvNode = pWeakEnvNode.lock();
+            if (pEnvNode)
+            {
+                pEnvNode->validate(status);
+            }
+        }
+    }
+
+
+    //
+    // Validate children, if so indicated
+    if (includeChildren)
+    {
+        for (auto &pSchemaChild: m_children)
+        {
+            pSchemaChild->validate(status, includeChildren, includeHiddenNodes);
+        }
+    }
+
 }

--- a/configuration/configmgr/configmgrlib/SchemaItem.hpp
+++ b/configuration/configmgr/configmgrlib/SchemaItem.hpp
@@ -28,6 +28,7 @@
 #include "platform.h"
 #include "EnvironmentEventHandlers.hpp"
 #include "ConfigPath.hpp"
+#include "Status.hpp"
 #include "Cfgmgrlib.hpp"
 
 class EnvironmentNode;
@@ -78,7 +79,7 @@ class CFGMGRLIB_API SchemaItem : public std::enable_shared_from_this<SchemaItem>
         void setProperty(const std::string &name, const std::string &value) { m_properties[name] = value; }
         void setHidden(bool hidden) { m_hidden = hidden; }
         bool isHidden() const { return m_hidden; }
-
+        std::shared_ptr<SchemaItem> getParent() const { return m_pParent.lock(); }
         void setParent(const std::shared_ptr<SchemaItem> &parent) { m_pParent = parent; }
         void setParentForChildren(std::shared_ptr<SchemaItem> pParent);
         std::shared_ptr<SchemaItem> getSchemaRoot();
@@ -87,6 +88,8 @@ class CFGMGRLIB_API SchemaItem : public std::enable_shared_from_this<SchemaItem>
         void getPath(std::string &path) const;
         void setRequiredInstanceComponents(const std::string list);
         const std::vector<std::string> &getRequiredInstanceComponents() const { return m_requiredInstanceComponents; }
+        void addEnvironmentNode(const std::shared_ptr<EnvironmentNode> &pEnvNode) { m_envNodes.emplace_back(pEnvNode); }
+        void validate(Status &status, bool includeChildren=false, bool includeHiddenNodes=false) const;
 
 
     protected:
@@ -124,6 +127,8 @@ class CFGMGRLIB_API SchemaItem : public std::enable_shared_from_this<SchemaItem>
 
         std::vector<std::shared_ptr<EnvironmentEventHandler>> m_eventHandlers;
         std::vector<std::string> m_requiredInstanceComponents;
+
+        mutable std::vector<std::weak_ptr<EnvironmentNode>> m_envNodes;
 
         // Following are NOT copied in copy constructor
         std::weak_ptr<SchemaItem> m_pParent;

--- a/configuration/configmgr/configmgrlib/XMLEnvironmentLoader.cpp
+++ b/configuration/configmgr/configmgrlib/XMLEnvironmentLoader.cpp
@@ -16,11 +16,10 @@
 ############################################################################## */
 
 #include "XMLEnvironmentLoader.hpp"
-#include "XSDSchemaParser.hpp"
 #include "Exceptions.hpp"
 
 
-std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentLoader::load(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem) const
+std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentLoader::load(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem, const std::string itemType) const
 {
     std::vector<std::shared_ptr<EnvironmentNode>> envNodes;
     std::shared_ptr<EnvironmentNode> pEnvNode;
@@ -29,7 +28,6 @@ std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentLoader::load(std::is
     try
     {
         pt::read_xml(in, envTree, pt::xml_parser::trim_whitespace | pt::xml_parser::no_comments);
-        //auto rootIt = envTree.begin();
 
         // if root, want to start with rootIt, but auto match rootIt first with schemaItem, then parse on down, but want to return
         // envNode for the root
@@ -37,6 +35,9 @@ std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentLoader::load(std::is
         // if not root, to through rootIt with schema item as parent for each elemement, return envNode of each item iterated
 
         std::shared_ptr<SchemaItem> pParseRootSchemaItem;
+
+        //
+        // For each top level element, parse it
         for (auto envIt = envTree.begin(); envIt != envTree.end(); ++envIt)
         {
             if (envIt->first == pSchemaItem->getProperty("name"))
@@ -46,12 +47,15 @@ std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentLoader::load(std::is
             else
             {
                 std::vector<std::shared_ptr<SchemaItem>> children;
-                pSchemaItem->getChildren(children, envIt->first);
+                pSchemaItem->getChildren(children, envIt->first, itemType);  // will only return 1 or no entries
                 if (children.empty())
                 {
                     throw (ParseException("Unable to start parsing environment, root node element " + envIt->first + " not found"));
                 }
-                pParseRootSchemaItem = children[0];
+                else
+                {
+                    pParseRootSchemaItem = children[0];
+                }
             }
 
             pEnvNode = std::make_shared<EnvironmentNode>(pParseRootSchemaItem, envIt->first);  // caller may need to set the parent
@@ -128,7 +132,10 @@ void XMLEnvironmentLoader::parse(const pt::ptree &envTree, const std::shared_ptr
             pConfigItem->getChildren(children, elemName, typeName);
             if (children.empty())
             {
-                pSchemaItem = std::make_shared<SchemaItem>(elemName, "default", pConfigItem);  // default item if none found
+                //
+                // No defined schema item found. Allocate a default one and add it to the tree
+                pSchemaItem = std::make_shared<SchemaItem>(elemName, "default", pConfigItem);
+                pConfigItem->addChild(pSchemaItem);
             }
             else if (children.size() > 1)
             {
@@ -139,8 +146,9 @@ void XMLEnvironmentLoader::parse(const pt::ptree &envTree, const std::shared_ptr
                 pSchemaItem = children[0];
             }
 
-            // If no schema item is found, that's ok, the node just has no defined configuration
+            // Note that if no schema item was found then a default schema item was allocated above and is used
             std::shared_ptr<EnvironmentNode> pElementNode = std::make_shared<EnvironmentNode>(pSchemaItem, elemName, pEnvNode);
+            pSchemaItem->addEnvironmentNode(pElementNode);
             parse(it->second, pSchemaItem, pElementNode);
             pEnvNode->addChild(pElementNode);
         }

--- a/configuration/configmgr/configmgrlib/XMLEnvironmentLoader.hpp
+++ b/configuration/configmgr/configmgrlib/XMLEnvironmentLoader.hpp
@@ -37,7 +37,8 @@ class CFGMGRLIB_API XMLEnvironmentLoader : public EnvironmentLoader
 
         XMLEnvironmentLoader() { }
         virtual ~XMLEnvironmentLoader() { }
-        virtual std::vector<std::shared_ptr<EnvironmentNode>> load(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem) const override;
+        virtual std::vector<std::shared_ptr<EnvironmentNode>> load(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem,
+                const std::string itemType = std::string("")) const override;
         void parse(const pt::ptree &envTree, const std::shared_ptr<SchemaItem> &pConfigItem, std::shared_ptr<EnvironmentNode> &pEnvNode) const;
 
 };

--- a/configuration/configmgr/configmgrlib/XMLEnvironmentMgr.cpp
+++ b/configuration/configmgr/configmgrlib/XMLEnvironmentMgr.cpp
@@ -29,13 +29,13 @@ bool XMLEnvironmentMgr::createParser()
 }
 
 
-std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentMgr::doLoadEnvironment(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem)
+std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentMgr::doLoadEnvironment(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem, const std::string itemType)
 {
     std::vector<std::shared_ptr<EnvironmentNode>> envNodes;
     try
     {
         XMLEnvironmentLoader envLoader;
-        envNodes = envLoader.load(in, pSchemaItem);
+        envNodes = envLoader.load(in, pSchemaItem, itemType);
     }
     catch (const std::exception &e)
     {
@@ -47,13 +47,30 @@ std::vector<std::shared_ptr<EnvironmentNode>> XMLEnvironmentMgr::doLoadEnvironme
 }
 
 
+bool XMLEnvironmentMgr::serialize(std::ostream &out, const std::shared_ptr<EnvironmentNode> &pStartNode)
+{
+    pt::ptree envTree, topTree;
+    serializeTree(envTree, pStartNode);
+    topTree.add_child(pStartNode->getName(), envTree);
+
+#if BOOST_VERSION >= 105800
+    pt::write_xml(out, topTree, pt::xml_parser::xml_writer_make_settings<std::string>(' ', 4));
+#else
+    const char * myIndent = " ";
+        pt::write_xml(out, topTree, pt::xml_parser::xml_writer_make_settings<char>(*myIndent , 4));
+#endif
+
+    return true;
+}
+
+
 bool XMLEnvironmentMgr::save(std::ostream &out)
 {
     bool rc = true;
     try
     {
         pt::ptree envTree, topTree;
-        serialize(envTree, m_pRootNode);
+        serializeTree(envTree, m_pRootNode);
         topTree.add_child("Environment", envTree);
 //        boost::property_tree::xml_writer_settings<std::string> settings;
         //pt::write_xml(out, topTree, pt::xml_parser::xml_writer_make_settings<std::string>(' ', 4));
@@ -75,7 +92,7 @@ bool XMLEnvironmentMgr::save(std::ostream &out)
 }
 
 
-void XMLEnvironmentMgr::serialize(pt::ptree &envTree, std::shared_ptr<EnvironmentNode> &pEnvNode) const
+void XMLEnvironmentMgr::serializeTree(pt::ptree &envTree, const std::shared_ptr<EnvironmentNode> &pEnvNode) const
 {
     std::vector<std::shared_ptr<EnvironmentValue>> attributes;
     pEnvNode->getAttributes(attributes);
@@ -99,7 +116,7 @@ void XMLEnvironmentMgr::serialize(pt::ptree &envTree, std::shared_ptr<Environmen
     for (auto childIt = children.begin(); childIt != children.end(); ++childIt)
     {
         pt::ptree nodeTree;
-        serialize(nodeTree, *childIt);
+        serializeTree(nodeTree, *childIt);
         envTree.add_child((*childIt)->getName(), nodeTree);
     }
 }

--- a/configuration/configmgr/configmgrlib/XMLEnvironmentMgr.hpp
+++ b/configuration/configmgr/configmgrlib/XMLEnvironmentMgr.hpp
@@ -32,14 +32,15 @@ class XMLEnvironmentMgr :   public EnvironmentMgr
 
         XMLEnvironmentMgr() : EnvironmentMgr() { }
         ~XMLEnvironmentMgr() { }
+        bool serialize(std::ostream &out, const std::shared_ptr<EnvironmentNode> &pStartNode);
 
 
     protected:
 
         bool createParser() override;
-        std::vector<std::shared_ptr<EnvironmentNode>> doLoadEnvironment(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem) override;
+        std::vector<std::shared_ptr<EnvironmentNode>> doLoadEnvironment(std::istream &in, const std::shared_ptr<SchemaItem> &pSchemaItem, const std::string itemType = std::string("")) override;
         bool save(std::ostream &out) override;
-        void serialize(pt::ptree &envTree, std::shared_ptr<EnvironmentNode> &pEnvNode) const;
+        void serializeTree(pt::ptree &envTree, const std::shared_ptr<EnvironmentNode> &pEnvNode) const;
 
 };
 

--- a/esp/scm/ws_configmgr.ecm
+++ b/esp/scm/ws_configmgr.ecm
@@ -286,6 +286,37 @@ ESPresponse [exceptions_inline] GetParentsResponse
 };
 
 
+ESPrequest GetNodeCopyRequest
+{
+    string SessionId;
+    string NodeId;
+};
+
+
+ESPresponse GetNodeCopyResponse
+{
+    string ValidInsertPath(""); // path to where the copy can be inserted
+    string NodeCopy("");        // opaque copy, DO NOT MODIFY
+};
+
+
+ESPrequest PasteNodeCopyRequest
+{
+    string SessionId;
+    string SessionLockKey;
+    string ParentNodeId;
+    string NodeCopy;
+};
+
+
+ESPresponse PasteNodeCopyResponse
+{
+    bool NodeInserted(false);
+    ESPstruct StatusType Status;
+    string NewNodeId("");
+};
+
+
 //
 // Environment management
 
@@ -484,6 +515,14 @@ ESPservice [auth_feature("DEFERRED"),version("2.0"), default_client_version("2.0
         [
             description("Fetch nodes matching the indicated path")
         ] FetchNodes(FetchNodesRequest, FetchNodesResponse);
+    ESPMethod
+        [
+            description("Get a copy of a node for paste operations")
+        ] GetNodeCopy(GetNodeCopyRequest, GetNodeCopyResponse);
+    ESPMethod
+        [
+            description("Paste the copy of a node to an environment")
+        ] InsertNodeCopy(PasteNodeCopyRequest, PasteNodeCopyResponse);
 };
 
 

--- a/esp/services/ws_configmgr/ws_configmgrError.hpp
+++ b/esp/services/ws_configmgr/ws_configmgrError.hpp
@@ -34,5 +34,6 @@
 #define CFGMGR_ERROR_ENVIRONMENT_LOCKING     CONFIG_MGR_ERROR_START+11   // Error locking the enironment
 #define CFGMGR_ERROR_NODE_INVALID            CONFIG_MGR_ERROR_START+12   // Enironment node not valid
 #define CFGMGR_ERROR_PATH_INVALID            CONFIG_MGR_ERROR_START+13   // The path specified is not valid
+#define CFGMGR_ERROR_NODE_COPY_INVALID       CONFIG_MGR_ERROR_START+14   // The node copy data is not valid
 
 #endif

--- a/esp/services/ws_configmgr/ws_configmgrService.cpp
+++ b/esp/services/ws_configmgr/ws_configmgrService.cpp
@@ -20,6 +20,7 @@
 #include "SchemaItem.hpp"
 #include "InsertableItem.hpp"
 #include "jexcept.hpp"
+#include "jutil.hpp"
 #include "ws_configmgrError.hpp"
 #include "Exceptions.hpp"
 
@@ -257,6 +258,8 @@ bool Cws_configMgrEx::onSaveEnvironmentFile(IEspContext &context, IEspSaveEnviro
         pSession = getConfigSessionForUpdate(req.getSessionLockKey(), req.getSessionLockKey());
     }
 
+    //
+    // Save it
     if (!pSession->saveEnvironment(saveFilename))
     {
         throw MakeStringException(CFGMGR_ERROR_SAVE_ENVIRONMENT, "There was a problem saving the environment");
@@ -269,11 +272,11 @@ bool Cws_configMgrEx::onSaveEnvironmentFile(IEspContext &context, IEspSaveEnviro
     // environment locked would have prevented this session from locking the same environment.
     if (!saveFilename.empty())
     {
-        for (auto sessionIt = m_sessions.begin(); sessionIt != m_sessions.end(); ++sessionIt)
+        for (auto &sessionIt: m_sessions)
         {
-            if (sessionIt->second->curEnvironmentFile == saveFilename)
+            if (sessionIt.second->curEnvironmentFile == saveFilename)
             {
-                sessionIt->second->externallyModified = true;
+                sessionIt.second->externallyModified = true;
             }
         }
     }
@@ -352,7 +355,7 @@ bool Cws_configMgrEx::onGetNode(IEspContext &context, IEspGetNodeRequest &req, I
     }
 
     getNodeResponse(pNode, resp);
-    pNode->validate(status, false);  // validate this node only
+    pNode->getSchemaItem()->validate(status, false);  // validate this node only
     buildStatusResponse(status, pSession, resp.updateStatus());
 
     //
@@ -587,6 +590,103 @@ bool Cws_configMgrEx::onFetchNodes(IEspContext &context, IEspFetchNodesRequest &
     catch (ParseException &pe)
     {
         throw MakeStringException(CFGMGR_ERROR_PATH_INVALID, "%s", pe.what());
+    }
+
+    return true;
+}
+
+
+bool Cws_configMgrEx::onGetNodeCopy(IEspContext &context, IEspGetNodeCopyRequest &req, IEspGetNodeCopyResponse &resp)
+{
+    std::string nodeId = req.getNodeId();
+    std::string sessionId = req.getSessionId();
+    Status status;
+
+    ConfigMgrSession *pSession = getConfigSession(sessionId, true);
+
+    if (nodeId.empty())
+        nodeId = pSession->m_pEnvMgr->getRootNodeId();
+
+    std::shared_ptr<EnvironmentNode> pNode = pSession->m_pEnvMgr->findEnvironmentNodeById(nodeId);
+    if (pNode)
+    {
+        std::ostringstream nodeCopy;
+        if (pSession->m_pEnvMgr->serialize(nodeCopy, pNode))
+        {
+            std::string path;
+            pNode->getSchemaItem()->getParent()->getPath(path);
+            std::string nodeBuffer = path + "|" + pNode->getSchemaItem()->getItemType() + "|" + nodeCopy.str();
+            StringBuffer encodedData;
+            JBASE64_Encode(nodeBuffer.c_str(), nodeBuffer.length(), encodedData, false);
+            resp.setNodeCopy(encodedData);
+            resp.setValidInsertPath(path.c_str());
+        }
+    }
+    else
+    {
+        throw MakeStringException(CFGMGR_ERROR_NODE_INVALID, "Environment node ID is not valid");
+    }
+    return true;
+}
+
+
+bool Cws_configMgrEx::onInsertNodeCopy(IEspContext &context, IEspPasteNodeCopyRequest &req, IEspPasteNodeCopyResponse &resp)
+{
+    std::string parentNodeId = req.getParentNodeId();
+    std::string sessionId = req.getSessionId();
+    Status status;
+
+    std::string key = req.getSessionLockKey();
+    ConfigMgrSession *pSession = getConfigSessionForUpdate(sessionId, key);
+
+    //
+    // Empty node means to paste at the root.
+    if (parentNodeId.empty())
+        parentNodeId = pSession->m_pEnvMgr->getRootNodeId();
+
+    std::shared_ptr<EnvironmentNode> pParentNode = pSession->m_pEnvMgr->findEnvironmentNodeById(parentNodeId);
+    if (pParentNode)
+    {
+        StringBuffer decodedDataBuffer;
+        JBASE64_Decode(req.getNodeCopy(), decodedDataBuffer);
+        std::string nodeData = decodedDataBuffer.str();
+
+        std::size_t barPos1 = nodeData.find_first_of('|', 0);
+        if (barPos1 == std::string::npos)
+        {
+            throw MakeStringException(CFGMGR_ERROR_NODE_INVALID, "The node copy data buffer is not valid");
+        }
+        std::size_t barPos2 = nodeData.find_first_of('|', barPos1+1);
+        if (barPos2 == std::string::npos)
+        {
+            throw MakeStringException(CFGMGR_ERROR_NODE_INVALID, "The node copy data buffer is not valid");
+        }
+        std::string path = nodeData.substr(0, barPos1);
+        std::string itemType = nodeData.substr(barPos1+1, barPos2 - barPos1 - 1);
+        std::string nodeCopy = nodeData.substr(barPos2+1);
+
+        std::shared_ptr <EnvironmentNode> pNewNode = pSession->m_pEnvMgr->addNewEnvironmentNode(pParentNode, nodeCopy,
+                                                                                                status, itemType);
+        //
+        // If the node was added, save the new ID and validate the environment. If there was an error, return an empty
+        // node ID. Caller cannot use the status to tell if node added because the validate may have errors. An empty
+        // new node ID indicates the node was not created.
+        if (pNewNode)
+        {
+            pSession->modified = true;
+            resp.setNewNodeId(pNewNode->getId().c_str());
+            resp.setNodeInserted(true);
+        }
+        else
+        {
+            resp.setNewNodeId("");
+        }
+
+        buildStatusResponse(status, pSession, resp.updateStatus());
+    }
+    else
+    {
+        throw MakeStringException(CFGMGR_ERROR_NODE_INVALID, "Environment node ID is not valid");
     }
 
     return true;

--- a/esp/services/ws_configmgr/ws_configmgrService.hpp
+++ b/esp/services/ws_configmgr/ws_configmgrService.hpp
@@ -56,6 +56,8 @@ public:
     virtual bool onGetOpenSessions(IEspContext &context, IEspListOpenSessionsRequest &req, IEspListOpenSessionsResponse &resp);
     virtual bool onGetNodeTree(IEspContext &context, IEspGetTreeRequest &req, IEspGetTreeResponse &resp);
     virtual bool onFetchNodes(IEspContext &context, IEspFetchNodesRequest &req, IEspFetchNodesResponse &resp);
+    virtual bool onGetNodeCopy(IEspContext &context, IEspGetNodeCopyRequest &req, IEspGetNodeCopyResponse &resp);
+    virtual bool onInsertNodeCopy(IEspContext &context, IEspPasteNodeCopyRequest &req, IEspPasteNodeCopyResponse &resp);
 
 
 private:

--- a/esp/services/ws_configmgr/ws_configmgrSession.hpp
+++ b/esp/services/ws_configmgr/ws_configmgrSession.hpp
@@ -120,7 +120,12 @@ struct ConfigMgrSession {
     {
         bool rc = false;
         std::string saveFile = (saveAsFilename != "") ? saveAsFilename : curEnvironmentFile;
-        if (m_pEnvMgr->saveEnvironment(saveFile))
+
+        std::ofstream out;
+        out.open(saveAsFilename);
+        std::shared_ptr<EnvironmentNode> pRootNode = m_pEnvMgr->findEnvironmentNodeById(m_pEnvMgr->getRootNodeId());
+
+        if (m_pEnvMgr->serialize(out, pRootNode))
         {
             modified = false;
             curEnvironmentFile = saveFile;

--- a/initfiles/componentfiles/configschema/xsd/dropzone.xsd
+++ b/initfiles/componentfiles/configschema/xsd/dropzone.xsd
@@ -20,7 +20,7 @@
         xmlns:hpcc="someuri">
     <xs:include schemaLocation="types.xsd"/>
     <hpcc:insert hpcc:schemaPath="/Environment/Software">
-        <xs:element name="DropZone" hpcc:displayName="Drop Zones" hpcc:itemType="dropzone" maxOccurs="unbounded"
+        <xs:element name="DropZone" hpcc:displayName="Drop Zones" hpcc:itemType="DropZone" maxOccurs="unbounded"
                     hpcc:docid="dz.t1" hpcc:class="component">
             <xs:complexType>
                 <xs:sequence>

--- a/initfiles/componentfiles/configschema/xsd/eclagent.xsd
+++ b/initfiles/componentfiles/configschema/xsd/eclagent.xsd
@@ -20,7 +20,7 @@
         xmlns:hpcc="someuri">
     <xs:include schemaLocation="types.xsd"/>
     <hpcc:insert hpcc:schemaPath="/Environment/Software">
-        <xs:element name="EclAgentProcess" hpcc:class="component" hpcc:itemType="eclAgent"
+        <xs:element name="EclAgentProcess" hpcc:class="component" hpcc:itemType="eclagent"
                     hpcc:displayName="ECL Agent Process" maxOccurs="unbounded" hpcc:docid="EA.t1">
             <xs:complexType>
                 <xs:attributeGroup name="options" hpcc:groupByName="Options" hpcc:docid="EA.t2">

--- a/initfiles/componentfiles/configschema/xsd/esp_service_desdl.xsd
+++ b/initfiles/componentfiles/configschema/xsd/esp_service_desdl.xsd
@@ -20,7 +20,7 @@
         xmlns:hpcc="someuri">
     <xs:include schemaLocation="types.xsd"/>
     <hpcc:insert hpcc:schemaPath="/Environment/Software">
-        <xs:element name="EspService" hpcc:class="component" hpcc:category="ESP Service" hpcc:itemType="dynesdl"
+        <xs:element name="EspService" hpcc:class="component" hpcc:category="ESP Service" hpcc:itemType="DynamicESDL"
                     hpcc:displayName="Dynamic ESDL" minOccurs="0" maxOccurs="unbounded">
             <xs:complexType>
                 <xs:sequence>

--- a/initfiles/componentfiles/configschema/xsd/esp_service_wssql.xsd
+++ b/initfiles/componentfiles/configschema/xsd/esp_service_wssql.xsd
@@ -20,7 +20,7 @@
         xmlns:hpcc="someuri">
     <xs:include schemaLocation="types.xsd"/>
     <hpcc:insert hpcc:schemaPath="/Environment/Software">
-        <xs:element name="EspService" hpcc:displayName="WsSQL Service" hpcc:itemType="wssql" minOccurs="0"
+        <xs:element name="EspService" hpcc:displayName="WsSQL Service" hpcc:itemType="ws_sql" minOccurs="0"
                     maxOccurs="unbounded" hpcc:class="component">
             <xs:complexType>
                 <xs:attributeGroup ref="buildInfo"/>

--- a/initfiles/componentfiles/configschema/xsd/logging_agent_cassandra.xsd
+++ b/initfiles/componentfiles/configschema/xsd/logging_agent_cassandra.xsd
@@ -21,7 +21,7 @@
     <xs:include schemaLocation="types.xsd"/>
     <hpcc:insert hpcc:schemaPath="/Environment/Software">
         <xs:element name="CassandraLoggingAgent" hpcc:displayName="Cassandra Logging Agent"
-                    hpcc:itemType="loggingagent_cassandra" hpcc:category="Logging Agent"
+                    hpcc:itemType="cassandraloggingagent" hpcc:category="Logging Agent"
                     maxOccurs="unbounded" hpcc:class="component">
             <xs:complexType>
                 <xs:attributeGroup name="cassandra" hpcc:groupByName="Cassandra">

--- a/initfiles/componentfiles/configschema/xsd/logging_agent_esp.xsd
+++ b/initfiles/componentfiles/configschema/xsd/logging_agent_esp.xsd
@@ -21,7 +21,7 @@
     <xs:include schemaLocation="types.xsd"/>
     <hpcc:insert hpcc:schemaPath="/Environment/Software">
         <xs:element name="ESPLoggingAgent" hpcc:displayName="ESP Logging Agent"
-                    hpcc:itemType="loggingagent_esp" hpcc:category="Logging Agent"
+                    hpcc:itemType="esploggingagent" hpcc:category="Logging Agent"
                     maxOccurs="unbounded" hpcc:class="component">
             <xs:complexType>
 

--- a/initfiles/componentfiles/configschema/xsd/logging_manager.xsd
+++ b/initfiles/componentfiles/configschema/xsd/logging_manager.xsd
@@ -20,7 +20,7 @@
         xmlns:hpcc="someuri">
     <xs:include schemaLocation="types.xsd"/>
     <hpcc:insert hpcc:schemaPath="/Environment/Software">
-        <xs:element name="LoggingManager" hpcc:displayName="Logging Manager" hpcc:itemType="logging_manager" minOccurs="0"
+        <xs:element name="LoggingManager" hpcc:displayName="Logging Manager" hpcc:itemType="loggingmanager" minOccurs="0"
                     maxOccurs="unbounded" hpcc:class="component">
             <xs:complexType>
                 <xs:sequence>

